### PR TITLE
 Waiting until prev block save,  Remove localnode when pick fetch node

### PR DESCRIPTION
**Pull request by anarcher**
### Background

- Waiting validator in sync until prev block is saved.
- Remove localNode when picking fetching node. 
- Validate block with prevblock's hash 

### Solution

- Using `common/observer` and `context.WithTimeout`
```
        ctx, cancelFunc := context.WithTimeout(pctx, v.prevBlockWaitTimeout)
	defer cancelFunc()

	prevHeight := height - 1

	exists, err := block.ExistsBlockByHeight(v.storage, prevHeight)
	if err != nil {
		return nil, err
	}

	if exists == false {
		waitC := make(chan struct{})
		observer.SyncBlockWaitObserver.On(string(prevHeight), func() {
			waitC <- struct{}{}
		})

		select {
		case <-waitC:
		case <-ctx.Done():
		}
	}
```

- Using `localNode.Address()` in `pickRandomNode`
- Make and hashing a block with local prev block and then compare this block's hash with fetched block's hash

### CODE REVIEW

File: lib/sync/fetcher.go [hunk 1]
```diff
@@ -86,18 +96,12 @@ func (f *BlockFetcher) fetch(ctx context.Context, si *SyncInfo) error {
 	f.logger.Debug("Fetch start", "height", height)
 
 	n := f.pickRandomNode()
-	f.logger.Info("Fetching items from node", "node", n, "height", height)
+	f.logger.Info(fmt.Sprintf("Fetching items from node: %s", n), "fetching_node", n, "height", height)
```
Comment (Charleslee522) [inline lib/sync/fetcher.go]: Isn't `n` duplicated?

File: lib/sync/fetcher.go [hunk 2]
```diff
@@ -175,8 +179,16 @@ func (f *BlockFetcher) pickRandomNode() node.Node {
 	if len(ac) <= 0 {
 		return nil
 	}
-	idx := rand.Intn(len(ac))
-	node := f.connectionManager.GetNode(ac[idx])
+
+	var addressList []string
+	for _, a := range ac {
+		if f.localNode.Address() != a {
```
Comment (Charleslee522) [inline lib/sync/fetcher.go]: :)

File: lib/sync/validator.go [hunk 3]
```diff
@@ -154,6 +160,15 @@ func (v *BlockValidator) validateBlock(ctx context.Context, si *SyncInfo) error
...
 		txs = append(txs, tx.H.Hash)
 	}
+
+	prevBlk, err := v.getPrevBlock(ctx, si.BlockHeight)
+	if err != nil {
+		return err
+	}
+
+	round := si.Block.Round
+	round.BlockHash = prevBlk.Hash
```
Comment (Charleslee522) [inline lib/sync/validator.go]: How about this?
`si.Block.Round.BlockHash = prevBlk.Hash`
First I saw `round` variable, it looked like unused :)

File: lib/sync/validator.go [hunk 4]
```diff
@@ -191,3 +206,34 @@ func (v *BlockValidator) existsBlock(ctx context.Context, st *storage.LevelDBBac
 		return exists, nil
 	}
 }
+
+func (v *BlockValidator) getPrevBlock(pctx context.Context, height uint64) (*block.Block, error) {
```
Comment (Charleslee522) [inline lib/sync/validator.go]: 👍

File: lib/sync/fetcher.go (hunk 1 above)
Comment (anarcher) [reply lib/sync/fetcher.go]: Yep, but msg has node is easy to see log :-)

**Comment by anarcher** [comment]
@Charleslee522  Can I merge it? ㅠㅠ

**Comment by anarcher** [comment]
Thanks! @Charleslee522 :-D

**Pull request merged**